tests/transforms/test_hamiltonian_expand.py

```diff
@@ -246,7 +247,252 @@ def test_hamiltonian_dif_tensorflow(self):
                     qml.CNOT(wires=[2, 0])
                 qml.expval(H)
 
-            tapes, fn = qml.transforms.hamiltonian_expand(tape)
+            tapes, fn = hamiltonian_expand(tape)
+            res = fn(qml.execute(tapes, dev, qml.gradients.param_shift, interface="tf"))
+
+            assert np.isclose(res, output)
+
+            g = gtape.gradient(res, var)
+            assert np.allclose(list(g[0]) + list(g[1]), output2)
+
+
+with QuantumTape() as s_tape1:
+    qml.PauliX(0)
+    S1 = qml.s_prod(1.5, qml.prod(qml.PauliZ(0), qml.PauliZ(1)))
+    qml.expval(S1)
+    qml.expval(S1)
+
+with QuantumTape() as s_tape2:
+    qml.Hadamard(0)
+    qml.Hadamard(1)
+    qml.PauliZ(1)
+    qml.PauliX(2)
+    S2 = qml.op_sum(
+        qml.prod(qml.PauliX(0), qml.PauliZ(2)),
+        qml.s_prod(3, qml.PauliZ(2)),
+        qml.s_prod(-2, qml.PauliX(0)),
+        qml.PauliX(2),
+        qml.prod(qml.PauliZ(0), qml.PauliX(1)),
+    )
+    qml.expval(S2)
+    qml.expval(S2)
+
+S3 = qml.op_sum(
+    qml.s_prod(1.5, qml.prod(qml.PauliZ(0), qml.PauliZ(1))), qml.s_prod(0.3, qml.PauliX(1))
+)
+
+with QuantumTape() as s_tape3:
+    qml.PauliX(0)
+    qml.expval(S3)
+    qml.expval(qml.PauliX(1))
+    qml.expval(S3)
```

Comment:
In this example where `S3` is computed twice, does it run the expectation value twice or is it clever enough to only do it once and make use of the results from the previous measurements? The latter functionality would be good (not just for having the same sums but also "overlapping" parts, i.e. `S1 = X1 X2 + Y1 Y2`, `S2 = X1 X2 + Z1 Z2` should only run 3 measurements, not 4.

Reply:
It runs the expectation twice... 😢 

Do you want me to add this functionality of re-using measurements?

Reply:
If its feasible for sure! But my impression is that this would go towards a bigger goal of taking _all_ obs of a return statement, and cleverly computing all of them in qwc groups. Probably beyond the scope of this here! And maybe something left for users to do by hand 🤔

Reply:
The individual tapes are created in this function! That is why we can do everything here.